### PR TITLE
Activate runtime PM support for i.MX8

### PR DESCRIPTION
**Pull request by dbaluta**
This PRs activates PM runtime support for OF device.

### CODE REVIEW

**Comment by dbaluta** [comment]
This was tested with real implementation of runtime/suspend. Will send those patches after this gets merged to make the review easier.

File: sound/soc/sof/sof-of-dev.c
Comment (plbossart) [inline sound/soc/sof/sof-of-dev.c]: are these last two needed?
Specifically the put_noidle was only needed for PCI, we don't use it anywhere else, do we?

File: sound/soc/sof/sof-of-dev.c
Comment (plbossart) [inline sound/soc/sof/sof-of-dev.c]: is this needed? It's the same as what snd_sof_prepare() does - or do you have CONFIG_ACPI enabled on i.MX?

Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.c]: For some reason we have CONFIG_ACPI enabled. I think we inherit it from generic arm64 defconfig.

File: sound/soc/sof/sof-of-dev.c
Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.c]: I will have a deeper look at this tommorrow. I've seen this pattern in multiple drivers. To be honest, I dont understand PM really well.

Comment (ranj063) [reply sound/soc/sof/sof-of-dev.c]: yup agree the last 2 shouldnt be needed and might even cause an imbalance. Just disabling PM in remove() should be good enough

File: sound/soc/sof/sof-of-dev.c
Comment (plbossart) [reply sound/soc/sof/sof-of-dev.c]: make we should have a helper that sets the value to S3 only, and a more fancy one that deals with the acpi_target_state().
@ranj063 what do you think?

Comment (ranj063) [reply sound/soc/sof/sof-of-dev.c]: @plbossart thats what I was thinking too but how do we decide which one to use? Should we base it on the platform config?

Comment (plbossart) [reply sound/soc/sof/sof-of-dev.c]: Not sure what you meant by platform config @ranj063 ? the point is that CONFIG_ACPI is not good enough, it's used in ARM platforms so can't be used to select Intel-specific code.

Comment (ranj063) [reply sound/soc/sof/sof-of-dev.c]: I meant the configs for each platform for ex: CONFIG_SND_SOC_SOF_BAYTRAIL and the others but that would be too much I think. Can we use these?
CONFIG_SND_SOC_SOF_PCI
CONFIG_SND_SOC_SOF_ACPI
CONFIG_SND_SOC_SOF_OF

Comment (lyakh) [reply sound/soc/sof/sof-of-dev.c]: Maybe improve granularity, in case of `snd_sof_prepare()` and S0IX support in general, add a check - run-time or compile-time - for whether it's supported?

Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.c]: @lyakh, initially I thought on adding a new field ```use_acpi``` in ```snd_sof_dev```. But I think your suggestion is better. 
Add a field ```pm_caps``` inside ```snd_sof_dev``` for supported PM states and deal with them in ```snd_sof_prepare```

**Comment by dbaluta** [comment]
@ranj063 @plbossart fixed my commits according to your comments. PR should OK to merge *after* we decide on how to deal with a generic snd_sof_prepare function.

**Review by plbossart — DISMISSED** [review]

File: sound/soc/sof/sof-of-dev.c
Comment (plbossart) [reply sound/soc/sof/sof-of-dev.c]: I don't think we want to hard-code power states, they might be dependent on firmware (BIOS) enabling specific states for e.g. PCI.

File: sound/soc/sof/sof-of-dev.c
Comment (kv2019i) [inline sound/soc/sof/sof-of-dev.c:61]: Hmm, it's a bit weird the sequence is different between OF and PCI, but this is probably the right thing to do and sof-pci-dev.c should call set_active() as well.

**Review by kv2019i — DISMISSED** [review]

File: sound/soc/sof/sof-of-dev.c
Comment (plbossart) [reply sound/soc/sof/sof-of-dev.c:61]: @kv2019i the PCI core already does this, see drivers/pci/pci.c

Comment (kv2019i) [reply sound/soc/sof/sof-of-dev.c:61]: @plbossart Aa, ok, thanks. I that makes sense. I was grepping through drivers, but redoing that, yeah, they are non-PCI devices.

**Comment by ranj063** [comment]
> @ranj063 @plbossart fixed my commits according to your comments. PR should OK to merge _after_ we decide on how to deal with a generic snd_sof_prepare function.

@dbaluta I see you've dropped the change for adding the special prepare callback for OF devices and there are 2 approvals already. I am confused about how it works for you?

**Comment by dbaluta** [comment]
@ranj063 as I said in a previous comment. We will merge this after we will find a solution for the common callback to work in all cases. Isn't this the idea?

Updated the PR title with DO NOT MERGE :)

**Comment by dbaluta** [comment]
@kv2019i @plbossart @ranj063 getting back at this, I've added a patch to fix the prepare callback for OF usecase. I think this is the cleanest solution we can have right now.

Another solution, would be for each driver (of, pci, acpi) to add their own implementations of prepare callback. In fact so far it looks like only of and pci drivers needs this function, sof-acpi driver doesn't register this callback for PM ops.

File: sound/soc/sof/pm.c
Comment (plbossart) [inline sound/soc/sof/pm.c]: !IS_ENABLED(CONFIG_SOF_OF) ?

Comment (plbossart) [reply sound/soc/sof/pm.c]: I don't think this works anyways, if someone enabled ACPI in your build it breaks your power management. too fragile.

I think we should add a boolean in the platform descriptors that defines if pm is managed by ACPI.

**Review by plbossart — CHANGES_REQUESTED** [review]
I don't think we can deal with this through build definitions, we need a platform-specific info on how PM is managed.

File: sound/soc/sof/pm.c
Comment (lyakh) [reply sound/soc/sof/pm.c]: @plbossart isn't the logic here exactly the opposite? As far as I understand this PR enables RT-PM on i:MX8, i.e. with `CONFIG_SOF_OF` enabled, right? Then according to this `#if` it doesn't matter whether `CONFIG_ACPI` is enabled - it anyway will take the `#else` branch

Comment (dbaluta) [reply sound/soc/sof/pm.c]: @lyakh @plbossart my intention is as @lyakh says. If CONFIG_SOF_OF is enabled then snd_sof_prepare will always select S3 by default as system_suspend_target. If CONFIG_SOF_OF is not enabled the behavior is unchanged.

Comment (dbaluta) [reply sound/soc/sof/pm.c]: @plbossart this can be an alternative as you suggested -> https://github.com/thesofproject/linux/pull/2441

**Comment by dbaluta** [comment]
@ranj063 @paulstelian97 @plbossart Now that https://github.com/thesofproject/linux/pull/2441 is merged we can continue adding PM support for i.MX8.

There are still two patches in this PR, one that activates runtime PM and the other one adds prepare/complete callbacks.

**Review by paulstelian97 — APPROVED** [review]
If it works, this looks good. Have you tested it?

**Comment by dbaluta** [comment]
@paulstelian97 yes, it works. I am always testing on linux-nxp/next tree and then apply the patches on topic/sof-dev. As you know i.MX8 boards support is not complete in upstream.

**Comment by dbaluta** [comment]
@ranj063 @plbossart @kv2019i you already reviewed this at the beginning of the PR, but the reviewed was discarded. 

Can you please help review it again?

@kv2019i I can help sending this to Mark on alsa-devel, as I noticed you are sending them.

**Review by lyakh — APPROVED** [review]

**Comment by lyakh** [comment]
@dbaluta maybe you could request a re-review from @plbossart ?

**Review by kv2019i — APPROVED** [review]
Looks good for me. @plbossart should be available today to do the review, I asked for a re-review. 
And @dbaluta , yes, thanks, feel free to send to Mark. The 5.10 window is closing, so I've been trying to empty our pile to Mark and Takashi and thus the many patches send this and last week.

**Review by plbossart — APPROVED** [review]
LGTM

**Pull request merged**
